packages/contracts/src/contracts/current/forwarder/MixinForwarderQuote.sol

```diff
@@ -0,0 +1,139 @@
+pragma solidity ^0.4.21;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+
+contract MixinForwarderQuote is MixinForwarderCore {
+
+    /// @dev Performs the 0x Exchange validation and calculations, without performing any state changes.
+    /// @param order An Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function fillOrderQuote(Order memory order, uint256 takerAssetFillAmount)
+        public
+        view
+        returns (Exchange.FillResults memory fillResults)
+    {
+        // Compute the order hash
+        bytes32 orderHash = getOrderHash(order);
+        uint256 remainingTakerAssetAmount = safeSub(order.takerAssetAmount, EXCHANGE.getUnavailableTakerTokenAmount(orderHash));
+        // Validate order expiration
+        if (block.timestamp >= order.expirationTimeSeconds) {
+            return fillResults;
+        }
+        // Validate order availability
+        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetAmount);
+        if (fillResults.takerAssetFilledAmount == 0) {
+            return fillResults;
+        }
+        // Validate fill order rounding
+        if (isRoundingError(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount)) {
+            fillResults.takerAssetFilledAmount = 0;
+            return fillResults;
+        }
+        fillResults.makerAssetFilledAmount = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount);
+        if (order.feeRecipientAddress != address(0)) {
+            if (order.makerFee > 0) {
+                fillResults.makerFeePaid = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerFee);
+            }
+            if (order.takerFee > 0) {
+                fillResults.takerFeePaid = getPartialAmount(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.takerFee);
+            }
+        }
+        return fillResults;
+    }
+
+    /// @dev Calculates a total for selling takerTokenFillAmount across all orders. Including the fees
+    ///      required to be paid. 
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function expectedMaketSellFillResults(Order[] memory orders, uint256 takerAssetFillAmount)
+        public
+        view
+        returns (Exchange.FillResults memory fillResult)
+    {
+        for (uint256 i = 0; i < orders.length; i++) {
+            require(areBytesEqual(orders[i].makerAssetData, orders[0].makerAssetData), "all orders must be the same token pair");
```

Comment:
This is only checking one side of the token pairing. Should we also be checking that the `takerAssetData` are the same?

Reply:
It will fail as the takerAssetData is always WETH and won't have a balance of any other tokens to  fill the order.